### ISSUE TITLE
VirtualService rules from ingress gateway to another or mesh does not appear to work

### ISSUE THREAD

**Issue by nathanshurte** [istio/istio#7369]
We've got a problem trying to apply routing rules via VirtualServices in the backend of our application. We have an ingress gateway with a virtual service, and these appear to work for inbound traffic, but for traffic internal (AFAICT the docs refer to this as "workflow") to to our backend services, I've been unable to get VirtualService rules applied. I've also been unable to get the second VirtualService to behave using the default `mesh` gateway either implicitly or explicitly. Likewise, DestinationRules don't appear to have an effect (we have no subsets currently).

Edit: this is on an Istio 1.0 snapshot

I've read through the docs thoroughly at this point and am either missing something obvious, have some hidden/broken config/infra preventing this from working, or this is a true bug/feature request.

We have the following ingress Gateway and VirtualService, which routes an 'gateway-service' which exposes access to our api, authentication services, etc., or defaults to a small proxy that provides appropriate frontend js based on DB lookup:

```
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: myapp-gateway
  namespace: myapp
spec:
  selector:
    istio: ingressgateway
  servers:
  - port:
      number: 80
      name: http
      protocol: HTTP
    hosts:
    - ui.qa.k8s.mydomain.com
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: myapp
  namespace: myapp
spec:
  hosts:
  - "ui.qa.k8s.mydomain.com"
  - myapp-ui
  - gateway-service
  gateways:
  - myapp-gateway
  http:
  - match:
    - uri:
        prefix: /api
    - uri:
        prefix: /uaa
    - uri:
        prefix: /admin
    route:
    - destination:
        host: gateway-host.myapp.svc.cluster.local # routes all to API facade/aggregtor, auth, etc. services
        port:
          number: 8080
  - route:
    - destination:
        host: ui-proxy.myapp.svc.cluster.local
        port:
          number: 80
```

The first VirtualService rules gets applied. When I go to create a new workflow via another Gateway + VirtualService (I've also tried via just a VirtualService using `mesh` as the gateway), these don't seem to work:

```
---
apiVersion: networking.istio.io./v1alpha3
kind: Gateway
metadata:
  name: prototype-backend-gateway
  namespace: myapp
spec:
  selector:
    track: release
  servers:
  - port:
      number: 8080
      name: http
      protocol: HTTP
    hosts:
    - target-service
```

Using, e.g., this virtualservice to slow all traffic:

```
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: slowresponse
  namespace: myapp
spec:
  gateways:
    - prototype-backend-gateway
  hosts:
  - target-service
  http:
  - route:
    - destination:
        host: target-service
        port:
          number: 8080
    fault:
      delay:
        percent: 100
        fixedDelay: 2s
```

In this case, the `target-service` is actually a few hops from the user: ui -> gateway -> api facade -> target-service.

tl;dr: We've been able to get get an ingress Gateway -> VirtualService -> actual service working, but can't seem to get ingress Gateway -> VirtualService -> actual service -> internal Gateway -> VirtualService -> backend service workflow established.

**Comment by rshriram** [comment]
where is the workload for the second gateway defined? (the one with `track: release` labels)

**Comment by blaketastic2** [comment]
@rshriram I'm working on this with @nathanshurte. What do you mean by "where is the workload"?

The thing is, that we've tried hitting the VirtualService with and without an internal gateway, with the assumption that it should be accessible from the "mesh".

**Comment by nathanshurte** [comment]
@rshriram My understanding is that creating the `prototype-backend-gateway` fronting `target-service` with the accompanying VirtualService `slowresponse` that specifies it uses that gateway and route into into `target-service` should create the workload. Is there something additional I'm missing?

**Comment by rshriram** [comment]
Gateway workloads need to be created explicitly. Take s look at helm charts on how we create ingress/egress gateways. 

Also short names are not allowed on the gateway

**Comment by Grant1219** [comment]
I am also running into this problem with Istio 1.0.2.

The `VirtualService` and `DestinationRule` seem to work fine when traffic comes through the ingress gateway, but when I removed the `gateway` block from my `VirtualService` YAML the rules and subset no longer apply, and traffic seems to simply go randomly to each deployment subset through the defined Kubernetes service.

How should clients inside the cluster send requests through "mesh-only" VirtualServices?

**Comment by blaketastic2** [comment]
@Grant1219 Can you add your VirtualService yaml?

**Comment by Grant1219** [comment]
@blaketastic2 Sure, here it is:
```yaml
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: k8s-stack-test-frontend-web
  namespace: application-platforms
spec:
  hosts:
    - "*"
  gateways:
    - mesh
  http:
  - route:
    - destination:
        host: k8s-stack-test-frontend-web
        port:
          number: 4567
        subset: v1
      weight: 90
    - destination:
        host: k8s-stack-test-frontend-web
        port:
          number: 4567
        subset: v2
      weight: 10

---
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: k8s-stack-test-frontend-web
  namespace: application-platforms
spec:
  host: k8s-stack-test-frontend-web
  trafficPolicy:
    loadBalancer:
      simple: RANDOM
  subsets:
  - name: v1
    labels:
      version: v1
  - name: v2
    labels:
      version: v2
```
I have tried several permutations, making weight for v1 "100", removing the v2 subset completely, and specifying the "hosts" in the `VirtualService` as `k8s-stack-frontend-gateway` instead of `"*"`, but nothing seems to take affect unless I attach the `VirtualService` to a `Gateway` and send traffic through the ingress at which point the rules and weights work as expected.

**Comment by blaketastic2** [comment]
Here is a working example I have that uses just mesh routing:
`
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: delayed-service
  namespace: foo
spec:
  hosts:
  - delayed-service.foo.svc.cluster.local
  gateways:
  - mesh
  http:
  - route:
    - destination:
        host: delayed-service.foo.svc.cluster.local
        port:
          number: 8080
    fault:
      delay:
        percent: 100
        fixedDelay: 30s
`
I couldn't quickly find an example of using the DestinationRule, but see if a simpler use case like the above works first.

The only things I can see that may need to change are:
- hosts "*" to k8s-stack-test-frontend-web.application-platforms.svc.cluster.local
- Reference all hosts with the fully qualified domain name(FQDN) as I mentioned above

If I come across an example with a DestinationRule today, I'll get that up here as well.

**Comment by Grant1219** [comment]
@blaketastic2 Thanks for those example YAMLs. It turns out specifying the FQDN actually solves the issue and routing rules within the mesh are now functional!

I did not test switching the protocol from http to tcp since the first example worked.

Do you have an idea of why the `VirtualService` works when attached to the `Gateway` only using the shortnames, but not when using the mesh gateway? I actually tried several variations and only the full domain (with `svc.cluster.local`) worked.

**Comment by blaketastic2** [comment]
@Grant1219 yea, I would open up an issue for that. We made the decision a while back to always use FQDN, and I will say that tightened things down, and cleared up a few issues we had way back at 0.6.

**Comment by blaketastic2** [comment]
@rshriram I know this is old, but we never got this working. Instead, we deployed 2 services and pointed to the service from our application. With that said, I'd love to get routing working in istio so we can point to 1 service, and have istio route to the correct deployment. I can also speak a bit more clearly as to what we are experiencing.

Currently, we route traffic from the default Istio IngressGateway to a VirtualService that is responsible for routing traffic either to our API or, as a catch all, to our UI via path based routing.
```yaml
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: app-gateway
  namespace: foo
spec:
  selector:
    istio: ingressgateway
  servers:
  - port:
      number: 80
      name: http
      protocol: HTTP
    hosts:
    - "*"
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: edb
  namespace: foo
spec:
  hosts:
  - "*"
  gateways:
  - app-gateway
  http:
  - match:
    - uri:
        prefix: /api
    route:
    - destination:
        host: api-gateway.foo.svc.cluster.local
        port:
          number: 8080
  - route:
    - destination:
        host: ui-router.foo.svc.cluster.local
        port:
          number: 80
```
Next, we use `ui-router` to populate a cookie(based on an API call it makes), and proxy the request along to our `ui` at it's FQDN `http://ui.foo.svc.cluster.local:80`. 

Because our `ui` would then need to determine which version it is routing to, we have defined a wrapping VirtualService:
```yaml
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: ui
  namespace: foo
spec:
  hosts:
  - ui.foo.svc.cluster.local
  gateways:
  - mesh
  http:
  - match:
    - headers:
        x-service-version: 
          exact: "2.5.0"
    route:
    - destination:
        host: ui.foo.svc.cluster.local
        subset: canary
  - route:
    - destination:
        host: ui.foo.svc.cluster.local
        subset: default
---
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: ui
  namespace: foo
spec:
  host: ui.foo.svc.cluster.local
  subsets:
  - name: default
    labels:
      version: default
  - name: canary
    labels:
      version: canary
```
At this point, our VirtualService named `ui` is not routing based on the header, but rather just appears to be doing either round robin or random to either deployment.

In a previous comment, you mentioned using a Gateway to solve this problem, so I've created another Gateway, `ui-gateway`, of type ClusterIP, that simply fronts our `ui` VirtualService. Is this correct so far, or was I misunderstanding you?

**Comment by emedina** [comment]
This is not working for me with `1.0.4`. I'm trying to set up a simple routing rule to the `reviews` service from the Bookinfo example:

```
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: reviews
  namespace: default
spec:
  host: reviews.default.svc.cluster.local
  subsets:
  - labels:
      version: v1
    name: v1
  - labels:
      version: v2
    name: v2
  - labels:
      version: v3
    name: v3
```
```
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: reviews
  namespace: default
spec:
  gateways:
  - mesh
  hosts:
  - reviews.default.svc.cluster.local
  http:
  - route:
    - destination:
        host: reviews.default.svc.cluster.local
        subset: v1
```

This configuration should always return `v1` of the `reviews` service, but it seems to be ignored as I just get a round-robin response between the three available versions:

```
# kubectl exec -ti sleep-7ffd5cc988-hhqvg -c sleep -- curl http://reviews.default.svc.cluster.local:9080/reviews/0
{"id": "0","reviews": [{  "reviewer": "Reviewer1",  "text": "An extremely entertaining play by Shakespeare. The slapstick humour is refreshing!", "rating": {"stars": 5, "color": "red"}},{  "reviewer": "Reviewer2",  "text": "Absolutely fun and entertaining. The play lacks thematic depth when compared to other plays by Shakespeare.", "rating": {"stars": 4, "color": "red"}}]}

# kubectl exec -ti sleep-7ffd5cc988-hhqvg -c sleep -- curl http://reviews.default.svc.cluster.local:9080/reviews/0
{"id": "0","reviews": [{  "reviewer": "Reviewer1",  "text": "An extremely entertaining play by Shakespeare. The slapstick humour is refreshing!", "rating": {"stars": 5, "color": "red"}},{  "reviewer": "Reviewer2",  "text": "Absolutely fun and entertaining. The play lacks thematic depth when compared to other plays by Shakespeare.", "rating": {"stars": 4, "color": "red"}}]}

# kubectl exec -ti sleep-7ffd5cc988-hhqvg -c sleep -- curl http://reviews.default.svc.cluster.local:9080/reviews/0
{"id": "0","reviews": [{  "reviewer": "Reviewer1",  "text": "An extremely entertaining play by Shakespeare. The slapstick humour is refreshing!", "rating": {"stars": 5, "color": "red"}},{  "reviewer": "Reviewer2",  "text": "Absolutely fun and entertaining. The play lacks thematic depth when compared to other plays by Shakespeare.", "rating": {"stars": 4, "color": "red"}}]}

# kubectl exec -ti sleep-7ffd5cc988-hhqvg -c sleep -- curl http://reviews.default.svc.cluster.local:9080/reviews/0
{"id": "0","reviews": [{  "reviewer": "Reviewer1",  "text": "An extremely entertaining play by Shakespeare. The slapstick humour is refreshing!", "rating": {"stars": 5, "color": "black"}},{  "reviewer": "Reviewer2",  "text": "Absolutely fun and entertaining. The play lacks thematic depth when compared to other plays by Shakespeare.", "rating": {"stars": 4, "color": "black"}}]}

# kubectl exec -ti sleep-7ffd5cc988-hhqvg -c sleep -- curl http://reviews.default.svc.cluster.local:9080/reviews/0
{"id": "0","reviews": [{  "reviewer": "Reviewer1",  "text": "An extremely entertaining play by Shakespeare. The slapstick humour is refreshing!", "rating": {"stars": 5, "color": "red"}},{  "reviewer": "Reviewer2",  "text": "Absolutely fun and entertaining. The play lacks thematic depth when compared to other plays by Shakespeare.", "rating": {"stars": 4, "color": "red"}}]}

# kubectl exec -ti sleep-7ffd5cc988-hhqvg -c sleep -- curl http://reviews.default.svc.cluster.local:9080/reviews/0
{"id": "0","reviews": [{  "reviewer": "Reviewer1",  "text": "An extremely entertaining play by Shakespeare. The slapstick humour is refreshing!", "rating": {"stars": 5, "color": "black"}},{  "reviewer": "Reviewer2",  "text": "Absolutely fun and entertaining. The play lacks thematic depth when compared to other plays by Shakespeare.", "rating": {"stars": 4, "color": "black"}}]}

# kubectl exec -ti sleep-7ffd5cc988-hhqvg -c sleep -- curl http://reviews.default.svc.cluster.local:9080/reviews/0
{"id": "0","reviews": [{  "reviewer": "Reviewer1",  "text": "An extremely entertaining play by Shakespeare. The slapstick humour is refreshing!"},{  "reviewer": "Reviewer2",  "text": "Absolutely fun and entertaining. The play lacks thematic depth when compared to other plays by Shakespeare."}]}

# kubectl exec -ti sleep-7ffd5cc988-hhqvg -c sleep -- curl http://reviews.default.svc.cluster.local:9080/reviews/0
{"id": "0","reviews": [{  "reviewer": "Reviewer1",  "text": "An extremely entertaining play by Shakespeare. The slapstick humour is refreshing!"},{  "reviewer": "Reviewer2",  "text": "Absolutely fun and entertaining. The play lacks thematic depth when compared to other plays by Shakespeare."}]}

# kubectl exec -ti sleep-7ffd5cc988-hhqvg -c sleep -- curl http://reviews.default.svc.cluster.local:9080/reviews/0
{"id": "0","reviews": [{  "reviewer": "Reviewer1",  "text": "An extremely entertaining play by Shakespeare. The slapstick humour is refreshing!", "rating": {"stars": 5, "color": "black"}},{  "reviewer": "Reviewer2",  "text": "Absolutely fun and entertaining. The play lacks thematic depth when compared to other plays by Shakespeare.", "rating": {"stars": 4, "color": "black"}}]}
```

I can confirm that the sleep pod I'm using has a working sidecar proxy injected:

```
# kubectl describe po sleep-7ffd5cc988-hhqvg
Name:               sleep-7ffd5cc988-hhqvg
Namespace:          default
Priority:           0
PriorityClassName:  <none>
Node:               orhlp-dck017/10.141.36.57
Start Time:         Mon, 10 Dec 2018 17:59:08 +0100
Labels:             app=sleep
                    pod-template-hash=3998177544
Annotations:        sidecar.istio.io/status={"version":"5bcec0bded3d8d85c82806404b076e86ff4987550efdbba37e2b9559f7cce7ee","initContainers":["istio-init"],"containers":["istio-proxy"],"volumes":["istio-envoy","istio-certs...
Status:             Running
IP:                 174.16.186.99
Controlled By:      ReplicaSet/sleep-7ffd5cc988
Init Containers:
  istio-init:
    Container ID:  docker://4fc410687ff982591aee9dd7b1df668608674dbcd4eed2831224efd5b940a55e
    Image:         docker.io/istio/proxy_init:1.0.4
    Image ID:      docker-pullable://istio/proxy_init@sha256:183e1a463846448c1ea7ecdb0fae0c94a523442254a0cde25ba12fa0bd17df0c
    Port:          <none>
    Host Port:     <none>
    Args:
      -p
      15001
      -u
      1337
      -m
      REDIRECT
      -i
      172.20.0.0/16
      -x

      -b

      -d

    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Mon, 10 Dec 2018 17:59:10 +0100
      Finished:     Mon, 10 Dec 2018 17:59:11 +0100
    Ready:          True
    Restart Count:  0
    Environment:    <none>
    Mounts:         <none>
```

I've also checked that the sleep pod has the `clusters` properly synced from the `pilot`:

```
# kubectl exec -ti sleep-7ffd5cc988-hhqvg -c istio-proxy -- curl http://localhost:15000/clusters | grep reviews
outbound|9080||reviews.default.svc.cluster.local::default_priority::max_connections::1024
outbound|9080||reviews.default.svc.cluster.local::default_priority::max_pending_requests::1024
outbound|9080||reviews.default.svc.cluster.local::default_priority::max_requests::1024
outbound|9080||reviews.default.svc.cluster.local::default_priority::max_retries::3
outbound|9080||reviews.default.svc.cluster.local::high_priority::max_connections::1024
outbound|9080||reviews.default.svc.cluster.local::high_priority::max_pending_requests::1024
outbound|9080||reviews.default.svc.cluster.local::high_priority::max_requests::1024
outbound|9080||reviews.default.svc.cluster.local::high_priority::max_retries::3
outbound|9080||reviews.default.svc.cluster.local::added_via_api::true
outbound|9080||reviews.default.svc.cluster.local::174.16.137.8:9080::cx_active::0
outbound|9080||reviews.default.svc.cluster.local::174.16.137.8:9080::cx_connect_fail::0
outbound|9080||reviews.default.svc.cluster.local::174.16.137.8:9080::cx_total::0
outbound|9080||reviews.default.svc.cluster.local::174.16.137.8:9080::rq_active::0
outbound|9080||reviews.default.svc.cluster.local::174.16.137.8:9080::rq_error::0
outbound|9080||reviews.default.svc.cluster.local::174.16.137.8:9080::rq_success::0
outbound|9080||reviews.default.svc.cluster.local::174.16.137.8:9080::rq_timeout::0
outbound|9080||reviews.default.svc.cluster.local::174.16.137.8:9080::rq_total::0
outbound|9080||reviews.default.svc.cluster.local::174.16.137.8:9080::health_flags::healthy
outbound|9080||reviews.default.svc.cluster.local::174.16.137.8:9080::weight::1
outbound|9080||reviews.default.svc.cluster.local::174.16.137.8:9080::region::
outbound|9080||reviews.default.svc.cluster.local::174.16.137.8:9080::zone::
outbound|9080||reviews.default.svc.cluster.local::174.16.137.8:9080::sub_zone::
outbound|9080||reviews.default.svc.cluster.local::174.16.137.8:9080::canary::false
outbound|9080||reviews.default.svc.cluster.local::174.16.137.8:9080::success_rate::-1
outbound|9080||reviews.default.svc.cluster.local::174.16.163.252:9080::cx_active::0
outbound|9080||reviews.default.svc.cluster.local::174.16.163.252:9080::cx_connect_fail::0
outbound|9080||reviews.default.svc.cluster.local::174.16.163.252:9080::cx_total::0
outbound|9080||reviews.default.svc.cluster.local::174.16.163.252:9080::rq_active::0
outbound|9080||reviews.default.svc.cluster.local::174.16.163.252:9080::rq_error::0
outbound|9080||reviews.default.svc.cluster.local::174.16.163.252:9080::rq_success::0
outbound|9080||reviews.default.svc.cluster.local::174.16.163.252:9080::rq_timeout::0
outbound|9080||reviews.default.svc.cluster.local::174.16.163.252:9080::rq_total::0
outbound|9080||reviews.default.svc.cluster.local::174.16.163.252:9080::health_flags::healthy
outbound|9080||reviews.default.svc.cluster.local::174.16.163.252:9080::weight::1
outbound|9080||reviews.default.svc.cluster.local::174.16.163.252:9080::region::
outbound|9080||reviews.default.svc.cluster.local::174.16.163.252:9080::zone::
outbound|9080||reviews.default.svc.cluster.local::174.16.163.252:9080::sub_zone::
outbound|9080||reviews.default.svc.cluster.local::174.16.163.252:9080::canary::false
outbound|9080||reviews.default.svc.cluster.local::174.16.163.252:9080::success_rate::-1
outbound|9080||reviews.default.svc.cluster.local::174.16.186.66:9080::cx_active::0
outbound|9080||reviews.default.svc.cluster.local::174.16.186.66:9080::cx_connect_fail::0
outbound|9080||reviews.default.svc.cluster.local::174.16.186.66:9080::cx_total::0
outbound|9080||reviews.default.svc.cluster.local::174.16.186.66:9080::rq_active::0
outbound|9080||reviews.default.svc.cluster.local::174.16.186.66:9080::rq_error::0
outbound|9080||reviews.default.svc.cluster.local::174.16.186.66:9080::rq_success::0
outbound|9080||reviews.default.svc.cluster.local::174.16.186.66:9080::rq_timeout::0
outbound|9080||reviews.default.svc.cluster.local::174.16.186.66:9080::rq_total::0
outbound|9080||reviews.default.svc.cluster.local::174.16.186.66:9080::health_flags::healthy
outbound|9080||reviews.default.svc.cluster.local::174.16.186.66:9080::weight::1
outbound|9080||reviews.default.svc.cluster.local::174.16.186.66:9080::region::
outbound|9080||reviews.default.svc.cluster.local::174.16.186.66:9080::zone::
outbound|9080||reviews.default.svc.cluster.local::174.16.186.66:9080::sub_zone::
outbound|9080||reviews.default.svc.cluster.local::174.16.186.66:9080::canary::false
outbound|9080||reviews.default.svc.cluster.local::174.16.186.66:9080::success_rate::-1
outbound|9080|v2|reviews.default.svc.cluster.local::default_priority::max_connections::1024
outbound|9080|v2|reviews.default.svc.cluster.local::default_priority::max_pending_requests::1024
outbound|9080|v2|reviews.default.svc.cluster.local::default_priority::max_requests::1024
outbound|9080|v2|reviews.default.svc.cluster.local::default_priority::max_retries::3
outbound|9080|v2|reviews.default.svc.cluster.local::high_priority::max_connections::1024
outbound|9080|v2|reviews.default.svc.cluster.local::high_priority::max_pending_requests::1024
outbound|9080|v2|reviews.default.svc.cluster.local::high_priority::max_requests::1024
outbound|9080|v2|reviews.default.svc.cluster.local::high_priority::max_retries::3
outbound|9080|v2|reviews.default.svc.cluster.local::added_via_api::true
outbound|9080|v2|reviews.default.svc.cluster.local::174.16.137.8:9080::cx_active::0
outbound|9080|v2|reviews.default.svc.cluster.local::174.16.137.8:9080::cx_connect_fail::0
outbound|9080|v2|reviews.default.svc.cluster.local::174.16.137.8:9080::cx_total::0
outbound|9080|v2|reviews.default.svc.cluster.local::174.16.137.8:9080::rq_active::0
outbound|9080|v2|reviews.default.svc.cluster.local::174.16.137.8:9080::rq_error::0
outbound|9080|v2|reviews.default.svc.cluster.local::174.16.137.8:9080::rq_success::0
outbound|9080|v2|reviews.default.svc.cluster.local::174.16.137.8:9080::rq_timeout::0
outbound|9080|v2|reviews.default.svc.cluster.local::174.16.137.8:9080::rq_total::0
outbound|9080|v2|reviews.default.svc.cluster.local::174.16.137.8:9080::health_flags::healthy
outbound|9080|v2|reviews.default.svc.cluster.local::174.16.137.8:9080::weight::1
outbound|9080|v2|reviews.default.svc.cluster.local::174.16.137.8:9080::region::
outbound|9080|v2|reviews.default.svc.cluster.local::174.16.137.8:9080::zone::
outbound|9080|v2|reviews.default.svc.cluster.local::174.16.137.8:9080::sub_zone::
outbound|9080|v2|reviews.default.svc.cluster.local::174.16.137.8:9080::canary::false
outbound|9080|v2|reviews.default.svc.cluster.local::174.16.137.8:9080::success_rate::-1
outbound|9080|v3|reviews.default.svc.cluster.local::default_priority::max_connections::1024
outbound|9080|v3|reviews.default.svc.cluster.local::default_priority::max_pending_requests::1024
outbound|9080|v3|reviews.default.svc.cluster.local::default_priority::max_requests::1024
outbound|9080|v3|reviews.default.svc.cluster.local::default_priority::max_retries::3
outbound|9080|v3|reviews.default.svc.cluster.local::high_priority::max_connections::1024
outbound|9080|v3|reviews.default.svc.cluster.local::high_priority::max_pending_requests::1024
outbound|9080|v3|reviews.default.svc.cluster.local::high_priority::max_requests::1024
outbound|9080|v3|reviews.default.svc.cluster.local::high_priority::max_retries::3
outbound|9080|v3|reviews.default.svc.cluster.local::added_via_api::true
outbound|9080|v3|reviews.default.svc.cluster.local::174.16.186.66:9080::cx_active::0
outbound|9080|v3|reviews.default.svc.cluster.local::174.16.186.66:9080::cx_connect_fail::0
outbound|9080|v3|reviews.default.svc.cluster.local::174.16.186.66:9080::cx_total::0
outbound|9080|v3|reviews.default.svc.cluster.local::174.16.186.66:9080::rq_active::0
outbound|9080|v3|reviews.default.svc.cluster.local::174.16.186.66:9080::rq_error::0
outbound|9080|v3|reviews.default.svc.cluster.local::174.16.186.66:9080::rq_success::0
outbound|9080|v3|reviews.default.svc.cluster.local::174.16.186.66:9080::rq_timeout::0
outbound|9080|v3|reviews.default.svc.cluster.local::174.16.186.66:9080::rq_total::0
outbound|9080|v3|reviews.default.svc.cluster.local::174.16.186.66:9080::health_flags::healthy
outbound|9080|v3|reviews.default.svc.cluster.local::174.16.186.66:9080::weight::1
outbound|9080|v3|reviews.default.svc.cluster.local::174.16.186.66:9080::region::
outbound|9080|v3|reviews.default.svc.cluster.local::174.16.186.66:9080::zone::
outbound|9080|v3|reviews.default.svc.cluster.local::174.16.186.66:9080::sub_zone::
outbound|9080|v3|reviews.default.svc.cluster.local::174.16.186.66:9080::canary::false
outbound|9080|v3|reviews.default.svc.cluster.local::174.16.186.66:9080::success_rate::-1
outbound|9080|v1|reviews.default.svc.cluster.local::default_priority::max_connections::1024
outbound|9080|v1|reviews.default.svc.cluster.local::default_priority::max_pending_requests::1024
outbound|9080|v1|reviews.default.svc.cluster.local::default_priority::max_requests::1024
outbound|9080|v1|reviews.default.svc.cluster.local::default_priority::max_retries::3
outbound|9080|v1|reviews.default.svc.cluster.local::high_priority::max_connections::1024
outbound|9080|v1|reviews.default.svc.cluster.local::high_priority::max_pending_requests::1024
outbound|9080|v1|reviews.default.svc.cluster.local::high_priority::max_requests::1024
outbound|9080|v1|reviews.default.svc.cluster.local::high_priority::max_retries::3
outbound|9080|v1|reviews.default.svc.cluster.local::added_via_api::true
outbound|9080|v1|reviews.default.svc.cluster.local::174.16.163.252:9080::cx_active::0
outbound|9080|v1|reviews.default.svc.cluster.local::174.16.163.252:9080::cx_connect_fail::0
outbound|9080|v1|reviews.default.svc.cluster.local::174.16.163.252:9080::cx_total::0
outbound|9080|v1|reviews.default.svc.cluster.local::174.16.163.252:9080::rq_active::0
outbound|9080|v1|reviews.default.svc.cluster.local::174.16.163.252:9080::rq_error::0
outbound|9080|v1|reviews.default.svc.cluster.local::174.16.163.252:9080::rq_success::0
outbound|9080|v1|reviews.default.svc.cluster.local::174.16.163.252:9080::rq_timeout::0
outbound|9080|v1|reviews.default.svc.cluster.local::174.16.163.252:9080::rq_total::0
outbound|9080|v1|reviews.default.svc.cluster.local::174.16.163.252:9080::health_flags::healthy
outbound|9080|v1|reviews.default.svc.cluster.local::174.16.163.252:9080::weight::1
outbound|9080|v1|reviews.default.svc.cluster.local::174.16.163.252:9080::region::
outbound|9080|v1|reviews.default.svc.cluster.local::174.16.163.252:9080::zone::
outbound|9080|v1|reviews.default.svc.cluster.local::174.16.163.252:9080::sub_zone::
outbound|9080|v1|reviews.default.svc.cluster.local::174.16.163.252:9080::canary::false
outbound|9080|v1|reviews.default.svc.cluster.local::174.16.163.252:9080::success_rate::-1
```

Although I see the ones without `version` prior to the ones with it (not sure if this is fine though).

Also I've checked the `routes`:

```
# ./istioctl pc routes sleep-7ffd5cc988-hhqvg -ojson
...
            {
                "name": "reviews.default.svc.cluster.local:9080",
                "domains": [
                    "reviews.default.svc.cluster.local",
                    "reviews.default.svc.cluster.local:9080",
                    "reviews",
                    "reviews:9080",
                    "reviews.default.svc.cluster",
                    "reviews.default.svc.cluster:9080",
                    "reviews.default.svc",
                    "reviews.default.svc:9080",
                    "reviews.default",
                    "reviews.default:9080",
                    "174.20.12.129",
                    "174.20.12.129:9080"
                ],
                "routes": [
                    {
                        "match": {
                            "prefix": "/"
                        },
                        "route": {
                            "cluster": "outbound|9080|v1|reviews.default.svc.cluster.local",
                            "timeout": "0.000s",
                            "maxGrpcTimeout": "0.000s"
                        },
                        "decorator": {
                            "operation": "reviews.default.svc.cluster.local:9080/*"
                        },
...
```

Any clue what can be going on here?

**Comment by darrencraig** [comment]
I had a similar issue to this which I've recently resolved. The system I was testing was a simple PHP service, fronted by an Nginx service. I've put the yaml files on GitHub so you can follow along: [https://github.com/darrencraig/istio-nginx-example](url)

The request process was as follows:

GET /example -> Istio Ingress -> Gateway -> Nginx Service & Virtual Service -> PHP Service / Virtual Service.

I couldn't get a VirtualService to bind to the PHP Service properly - it appeared as if the routing rules were being ignored.

The core of my problem lay a silly detail - the connection between Nginx and the PHP services was TCP, not HTTP. I had previously defined http bindings in my Yaml, but I had to change the type of connection for my PHP Virtual Service to TCP - e.g. 

[https://github.com/darrencraig/istio-nginx-example/blob/master/kubernetes/ingress.yaml#L46](url)

This worked, but created a second problem for me - you cannot split/weight traffic on a TCP connection in Istio 1.0.x. I did a bit of investigating and found that support has been added for TCP weighting, and should be released with version 1.1 in January.

I hope that helps someone!

**Comment by emedina** [comment]
@darrencraig Changing `http` to `tcp` didn't make any change on my side. The rule is completely ignored still:

```
# for i in `seq 4`; do   kubectl exec -it $SLEEP_POD -c sleep -- curl http://reviews:9080/reviews/0 | jq '.reviews|.[]|.rating?'; done
{
  "stars": 5,
  "color": "red"
}
{
  "stars": 4,
  "color": "red"
}
null
null
{
  "stars": 5,
  "color": "black"
}
{
  "stars": 4,
  "color": "black"
}
null
null
```

**Comment by je-al** [comment]
FWIW, I thought I was experiencing this issue but it turned out to be I hadn't noticed naming the port exposed in the service (regular k8s service) is *mandatory*, i.e.:
```
apiVersion: v1
kind: Service
...
spec:
  ports:
  - name: http
    port: 9080
    protocol: TCP
    targetPort: 9080
...
```

**Comment by jpreese** [comment]
+1. Chiming in to say that port name is required for VS rules to apply.
